daemon/cmd/daemon.go

```diff
@@ -460,7 +460,12 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 
 	d.k8sWatcher.NodeChain.Register(d.endpointManager)
 	if option.Config.BGPAnnounceLBIP || option.Config.BGPAnnouncePodCIDR {
-		d.k8sWatcher.NodeChain.Register(d.bgpSpeaker)
+		switch option.Config.IPAMMode() {
+		case ipamOption.IPAMKubernetes:
+			d.k8sWatcher.NodeChain.Register(d.bgpSpeaker)
+		case ipamOption.IPAMClusterPool:
+			d.k8sWatcher.CiliumNodeChain.Register(d.bgpSpeaker)
```

Comment:
Are these events invoked in the same fashion as the Node Events? Will be reliably get "NodeDelete" events for CiliumNode objects as well?

Reply:
Are there specific requirements around deletion? Generally speaking, we do assume that CiliumNode objects are deleted eventually, since otherwise the pod CIDRs assigned to them won't be available to any other nodes. There is an outstanding bug regarding this: https://github.com/cilium/cilium/issues/17837 - but in principle we rely on CiliumNode objects getting deleted for IPAM correctness too.

Overall, I'm not sure how timely these deletion events will occur. I assume what will happen in practice is that the pod CIDR routes to stale nodes are kept around until CRD garbage collection kicks in. However, since the pod CIDR (from which the routes are derived) will also not be released while the CRD is alive, there should be no danger of conflicting routes.

Reply:
It's a pretty soft requirement TBH. For full context:

The BGP code will see a NodeDelete object and attempt a graceful withdrawal of BGP routes from its peer. 

However, if the node goes off line, the peer will understand that and perform its own withdrawal of the routes. 

So, given your comments, I think we are OK either way. Its just that for graceful withdrawal we'd want to see CiliumNode delete events when the node is being taken out of the cluster.

Reply:
FWIW, CiliumNode deletion relies on the ownerReference to be set in the CiliumNode resource pointing to the corresponding Node. In other words, when an ownerReference is set, the deletion of the referred will also delete the referee.